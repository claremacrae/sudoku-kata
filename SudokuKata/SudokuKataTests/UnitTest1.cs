using ApprovalTests;
using ApprovalTests.Reporters;
using SudokuKata;
using System;
using System.IO;
using Xunit;

namespace SudokuKataTests
{
    public class UnitTest1
    {

        // From http://www.vtrifonov.com/2012/11/getting-console-output-within-unit-test.html
        public class ConsoleOutput : IDisposable
        {
            private StringWriter stringWriter;
            private TextWriter originalOutput;

            public ConsoleOutput()
            {
                stringWriter = new StringWriter();
                originalOutput = Console.Out;
                Console.SetOut(stringWriter);
            }

            public string GetOuput()
            {
                return stringWriter.ToString();
            }

            public void Dispose()
            {
                Console.SetOut(originalOutput);
                stringWriter.Dispose();
            }
        }

 
        [UseReporter(typeof(AraxisMergeReporter))]
        [Fact]
        public void Test1()
        {
            VerifySudokuForSeed(42);
        }
        
        [UseReporter(typeof(AraxisMergeReporter))]
        [Fact]
        public void Test2()
        {
            VerifySudokuForSeed(1);
        }

        private static void VerifySudokuForSeed(int seed)
        {
            var currentConsoleOut = Console.Out;
            using (var consoleOutput = new ConsoleOutput())
            {
                SudokuKata.Program.Play(new Random(seed));
                string s = consoleOutput.GetOuput();
                Approvals.Verify(s);
            }
            Assert.Equal(currentConsoleOut, Console.Out);
        }
    }
}
