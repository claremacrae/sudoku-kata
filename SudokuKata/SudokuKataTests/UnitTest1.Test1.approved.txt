
Final look of the solved board:
+---+---+---+
|947|253|618|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|295|731|846|
|784|526|931|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+

Starting look of the board to solve:
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|..5|
+---+---+---+
|6..|49.|.7.|
|...|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|8..|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+

================================================================================

Column #1 can contain 2 only at (5, 1).
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|..5|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|8..|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040200000520004097806007005600490070200000006704006030002800000300000709079305200

Block (1, 3) can contain 2 only at (3, 8).
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|.25|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|8..|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040200000520004097806007025600490070200000006704006030002800000300000709079305200

Row #7 can contain 7 only at (7, 5).
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|.25|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|87.|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040200000520004097806007025600490070200000006704006030002870000300000709079305200

Block (1, 3) can contain 4 only at (3, 7).
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|87.|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040200000520004097806007425600490070200000006704006030002870000300000709079305200

Block (3, 2) can contain 9 only at (7, 6).
+---+---+---+
|.4.|2..|...|
|52.|..4|.97|
|8.6|..7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040200000520004097806007425600490070200000006704006030002879000300000709079305200

Row #1 can contain 5 only at (1, 5).
+---+---+---+
|.4.|25.|...|
|52.|..4|.97|
|8.6|..7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040250000520004097806007425600490070200000006704006030002879000300000709079305200

Block (1, 2) can contain 9 only at (3, 4).
+---+---+---+
|.4.|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 040250000520004097806907425600490070200000006704006030002879000300000709079305200

Row #1 can contain 7 only at (1, 3).
+---+---+---+
|.47|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|..6|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 047250000520004097806907425600490070200000006704006030002879000300000709079305200

Block (2, 3) can contain 4 only at (5, 8).
+---+---+---+
|.47|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|.46|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|...|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 047250000520004097806907425600490070200000046704006030002879000300000709079305200

Row #8 can contain 4 only at (8, 5).
+---+---+---+
|.47|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|.46|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 047250000520004097806907425600490070200000046704006030002879000300040709079305200

Block (1, 1) can contain 9 only at (1, 1).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|.46|
|7.4|..6|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490070200000046704006030002879000300040709079305200

Column #5 can contain 2 only at (6, 5).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.7.|
|2..|...|.46|
|7.4|.26|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490070200000046704026030002879000300040709079305200

Row #4 can contain 2 only at (4, 9).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|...|.46|
|7.4|.26|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490072200000046704026030002879000300040709079305200

Column #4 can contain 7 only at (5, 4).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|.26|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490072200700046704026030002879000300040709079305200

Block (2, 2) can contain 5 only at (6, 4).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|.4.|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490072200700046704526030002879000300040709079305200

Row #8 can contain 2 only at (8, 6).
+---+---+---+
|947|25.|...|
|52.|..4|.97|
|8.6|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|.42|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000520004097806907425600490072200700046704526030002879000300042709079305200

Values 1 and 4 in block (3, 1) are in cells (7, 1) and (9, 1).
1 cannot appear in (7, 2).
1 cannot appear in (8, 2).
1 cannot appear in (8, 3).
Guessing that 1 and 3 are arbitrary in block (1, 1) (multiple solutions): Pick 1->(2, 3), 3->(3, 2).
+---+---+---+
|947|25.|...|
|521|..4|.97|
|836|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|.42|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000521004097836907425600490072200700046704526030002879000300042709079305200

(2, 4) can only contain 6.
+---+---+---+
|947|25.|...|
|521|6.4|.97|
|836|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|.42|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000521604097836907425600490072200700046704526030002879000300042709079305200

(8, 4) can only contain 1.
+---+---+---+
|947|25.|...|
|521|6.4|.97|
|836|9.7|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000521604097836907425600490072200700046704526030002879000300142709079305200

(3, 5) can only contain 1.
+---+---+---+
|947|25.|...|
|521|6.4|.97|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|3.5|2..|
+---+---+---+
Code: 947250000521604097836917425600490072200700046704526030002879000300142709079305200

(9, 5) can only contain 6.
+---+---+---+
|947|25.|...|
|521|6.4|.97|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|365|2..|
+---+---+---+
Code: 947250000521604097836917425600490072200700046704526030002879000300142709079365200

Values 1 and 4 in block (3, 1) are in cells (7, 1) and (9, 1).
1 cannot appear in (7, 2).
Guessing that 3 and 8 are arbitrary in block (1, 2) (multiple solutions): Pick 3->(1, 6), 8->(2, 5).
+---+---+---+
|947|253|...|
|521|684|.97|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|365|2..|
+---+---+---+
Code: 947253000521684097836917425600490072200700046704526030002879000300142709079365200

(2, 7) can only contain 3.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|7..|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|365|2..|
+---+---+---+
Code: 947253000521684397836917425600490072200700046704526030002879000300142709079365200

(5, 5) can only contain 3.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|...|
|3..|142|7.9|
|.79|365|2..|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030002879000300142709079365200

Column #9 can contain 3 only at (7, 9).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|..3|
|3..|142|7.9|
|.79|365|2..|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030002879003300142709079365200

Block (3, 3) can contain 4 only at (9, 9).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|..3|
|3..|142|7.9|
|.79|365|2.4|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030002879003300142709079365204

(9, 1) can only contain 1.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|..3|
|3..|142|7.9|
|179|365|2.4|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030002879003300142709179365204

(9, 8) can only contain 8.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|..2|879|..3|
|3..|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030002879003300142709179365284

(7, 1) can only contain 4.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6..|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|4.2|879|..3|
|3..|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425600490072200730046704526030402879003300142709179365284

Row #4 can contain 3 only at (4, 3).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|.72|
|2..|73.|.46|
|7.4|526|.3.|
+---+---+---+
|4.2|879|..3|
|3..|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425603490072200730046704526030402879003300142709179365284

Guessing that 5 and 8 are arbitrary in column #3 (multiple solutions): Pick 5->(5, 3), 8->(8, 3).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|.72|
|2.5|73.|.46|
|7.4|526|.3.|
+---+---+---+
|4.2|879|..3|
|3.8|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425603490072205730046704526030402879003308142709179365284

Block (2, 3) can contain 5 only at (4, 7).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|572|
|2.5|73.|.46|
|7.4|526|.3.|
+---+---+---+
|4.2|879|..3|
|3.8|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425603490572205730046704526030402879003308142709179365284

Guessing that 5 and 6 are arbitrary in column #2 (multiple solutions): Pick 5->(7, 2), 6->(8, 2).
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|572|
|2.5|73.|.46|
|7.4|526|.3.|
+---+---+---+
|452|879|..3|
|368|142|7.9|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425603490572205730046704526030452879003368142709179365284

(8, 8) can only contain 5.
+---+---+---+
|947|253|...|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|572|
|2.5|73.|.46|
|7.4|526|.3.|
+---+---+---+
|452|879|..3|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253000521684397836917425603490572205730046704526030452879003368142759179365284

Guessing that 1 and 8 are arbitrary in column #9 (multiple solutions): Pick 8->(1, 9), 1->(6, 9).
+---+---+---+
|947|253|..8|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|49.|572|
|2.5|73.|.46|
|7.4|526|.31|
+---+---+---+
|452|879|..3|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253008521684397836917425603490572205730046704526031452879003368142759179365284

Guessing that 1 and 8 are arbitrary in column #6 (multiple solutions): Pick 8->(4, 6), 1->(5, 6).
+---+---+---+
|947|253|..8|
|521|684|397|
|836|917|425|
+---+---+---+
|6.3|498|572|
|2.5|731|.46|
|7.4|526|.31|
+---+---+---+
|452|879|..3|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253008521684397836917425603498572205731046704526031452879003368142759179365284

(4, 2) can only contain 1.
+---+---+---+
|947|253|..8|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|.46|
|7.4|526|.31|
+---+---+---+
|452|879|..3|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253008521684397836917425613498572205731046704526031452879003368142759179365284

Guessing that 1 and 6 are arbitrary in row #7 (multiple solutions): Pick 1->(7, 7), 6->(7, 8).
+---+---+---+
|947|253|..8|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|.46|
|7.4|526|.31|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253008521684397836917425613498572205731046704526031452879163368142759179365284

(1, 7) can only contain 6.
+---+---+---+
|947|253|6.8|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|.46|
|7.4|526|.31|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253608521684397836917425613498572205731046704526031452879163368142759179365284

(1, 8) can only contain 1.
+---+---+---+
|947|253|618|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|.46|
|7.4|526|.31|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253618521684397836917425613498572205731046704526031452879163368142759179365284

Guessing that 8 and 9 are arbitrary in row #6 (multiple solutions): Pick 8->(6, 2), 9->(6, 7).
+---+---+---+
|947|253|618|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|.46|
|784|526|931|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253618521684397836917425613498572205731046784526931452879163368142759179365284

(5, 7) can only contain 8.
+---+---+---+
|947|253|618|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|2.5|731|846|
|784|526|931|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253618521684397836917425613498572205731846784526931452879163368142759179365284

(5, 2) can only contain 9.
+---+---+---+
|947|253|618|
|521|684|397|
|836|917|425|
+---+---+---+
|613|498|572|
|295|731|846|
|784|526|931|
+---+---+---+
|452|879|163|
|368|142|759|
|179|365|284|
+---+---+---+
Code: 947253618521684397836917425613498572295731846784526931452879163368142759179365284

